### PR TITLE
Oracle Rate Generator and Oracle Provider

### PR DESCRIPTION
**Pull request by sissnad**
This PR provides:
- an `OraclePriceGenerator` class
- an `OracleProvider` class

It additionally fixes some issues with the method `historical_returns()` in `MarketPriceGenerator`

### CODE REVIEW

File: model/generators/markets.py [hunk 1]
```diff
@@ -57,44 +58,34 @@ def __init__(
...
-            "celo": np.zeros(
-                simulation_configuration.BLOCKS_PER_TIMESTEP
-                * simulation_configuration.TIMESTEPS
-                + 1
-            ),
-        }
+        self.supply_changes = {asset.split('_')[0]: np.zeros(
+            simulation_configuration.BLOCKS_PER_TIMESTEP
+            * simulation_configuration.TIMESTEPS
+            + 1) for asset in impacted_assets}
```
Comment (tobikuhlmann) [inline model/generators/markets.py:65]: nice :)

File: model/generators/oracles.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+ACCOUNTS_NS = uuid4()
+
+# raise numpy warnings as errors
+np.seterr(all='raise')
+
+
+class OracleRateGenerator(Generator):
+    """
+    This class is providing a market environment
```
Comment (tobikuhlmann) [inline model/generators/oracles.py:23]: is this description correct?

Comment (tobikuhlmann) [inline model/generators/oracles.py:23]: is this description correct?

File: model/parts/market_prices.py [hunk 3]
```diff
@@ -19,23 +19,20 @@ def p_market_price(
...
-    }
     return {
         "market_price": market_price,
-        "market_buckets": market_buckets,
-        "oracle_rate": market_price["celo_usd"],
     }
 
 
-@inject(MarketPriceGenerator)
+@ inject(MarketPriceGenerator)
```
Comment (tobikuhlmann) [inline model/parts/market_prices.py:27]: should the space be removed here?

Comment (tobikuhlmann) [inline model/parts/market_prices.py:27]: should the space be removed here?

File: model/parts/market_prices.py [hunk 4]
```diff
@@ -19,23 +19,20 @@ def p_market_price(
...
 def p_price_impact(
     params,
     _substep,
     state_history,
     prev_state,
     market_price_generator: MarketPriceGenerator,
+
+
```
Comment (tobikuhlmann) [inline model/parts/market_prices.py:35]: let's remove these empty lines

Comment (tobikuhlmann) [inline model/parts/market_prices.py:35]: let's remove these empty lines

File: model/parts/policies_oracles.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    params,
+    _substep,
+    state_history,
+    prev_state,
+    oracle_rate_generator: OracleRateGenerator,
+):
+    """
+    Updates the median rate to update the Mento buckets
+    """
+    if (prev_state['timestep'] > 1) and buckets_should_be_reset(params, prev_state):
```
Comment (tobikuhlmann) [inline model/parts/policies_oracles.py:21]: I'm not sure I understand this. Why only update the oracle rate on a bucket update, and not every block?

Comment (tobikuhlmann) [inline model/parts/policies_oracles.py:21]: I'm not sure I understand this. Why only update the oracle rate on a bucket update, and not every block?

Comment (tobikuhlmann) [reply model/parts/policies_oracles.py:21]: is it only for saving computation? I think validator rewards are mined at oracle rate every epoch block, but not sure if this is significant whether it uses an updated oracle rate or from 1 bucket update behind

Comment (tobikuhlmann) [reply model/parts/policies_oracles.py:21]: is it only for saving computation? I think validator rewards are mined at oracle rate every epoch block, but not sure if this is significant whether it uses an updated oracle rate or from 1 bucket update behind

File: model/utils/data_feed.py [hunk 6]
```diff
@@ -5,17 +5,24 @@
 from pathlib import Path
 import numpy as np
 import pandas as pd
-import data.mock_data # this is necessary to create mock data if not existent
+# pylint: disable = unused-import
+# pylint: disable = redefined-outer-name
```
Comment (tobikuhlmann) [inline model/utils/data_feed.py:9]: nice, this is probably cleaner than committing with `--no-verify`

Comment (tobikuhlmann) [inline model/utils/data_feed.py:9]: nice, this is probably cleaner than committing with `--no-verify`

**Review by tobikuhlmann — COMMENTED** [review]
I like the structure :) and left a couple of comments where I could need clarification

File: model/generators/oracles.py (hunk 2 above)
Comment (sissnad) [reply model/generators/oracles.py]: Changed!

File: model/parts/market_prices.py (hunk 3 above)
Comment (sissnad) [reply model/parts/market_prices.py]: Done!

File: model/generators/oracles.py [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+                                         oracle_id=uuid5(ACCOUNTS_NS, oracle_name),
+                                         aggregation=aggregation, delay=10
+                                         )
+        self.oracles_by_id[oracle_provider.oracle_id] = oracle_provider
+        return oracle_provider
+
+    def delay(self):
+        """
+        Delay function for advanced delaying or different delay types (e.r. random)
+        """
```
Comment (bowd) [inline model/generators/oracles.py:57]: What is this for?

File: model/parts/policies_oracles.py (hunk 5 above)
Comment (bowd) [reply model/parts/policies_oracles.py:21]: +1 not sure what the intent is here, the oracles should have their own report rate, right?

File: model/generators/oracles.py [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+
+    def exchange_rate(self, state_history, prev_state):
+        exchange_rate = self.aggregation(state_history, prev_state)
+        return exchange_rate
+
+    def aggregation(self, state_history, prev_state):
+        oracle_reports = [oracle.report(state_history, prev_state)
+                          for _oracle_id, oracle in self.oracles_by_id.items()]
+        return oracle_reports[0]
```
Comment (bowd) [inline model/generators/oracles.py:66]: Two points:
1. I'll refactor this in my indie mento branch to make oracles exist for all pairs that we're interested in and then the aggregation here should happen by pair, but it's fine to treat this separately.
2. What's the intent of the aggregation here, it looks like we're just taking the first report.

File: model/parts/policies_oracles.py (hunk 5 above)
Comment (sissnad) [reply model/parts/policies_oracles.py]: I have changed it. Check out [`update()`](https://github.com/celo-org/mento2-model/blob/a77154022fc57b670a779924312b8cf725a7b8a4/model/entities/oracle_provider.py#L25)

Comment (sissnad) [reply model/parts/policies_oracles.py]: See above

File: model/entities/oracle_provider.py [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+        self.oracle_id = oracle_id
+        self.aggregation_method = aggregation
+        self.delay = delay
+        self.reports = {'celo_usd': None}
+        self.rng = rngp.get_rng("Oracle")
+
+    def aggregation(self, _state_history, prev_state):
+        if isinstance(self.aggregation_method, AggregationMethod):
+            oracle_report = prev_state["market_price"]['celo_usd']
+        self.reports['celo_usd'] = np.concatenate((self.reports, oracle_report))
```
[the remark this answers is not in the document]
Comment (sissnad) [reply model/entities/oracle_provider.py]: Removed for now. My idea was to have a structure for future more general aggregation

File: model/generators/oracles.py (hunk 7 above)
Comment (sissnad) [reply model/generators/oracles.py]: Removed

**Comment by sissnad** [comment]
@bowd I think I have addressed all points. I have also merged main into this branch.

**Review by bowd — APPROVED** [review]

**Pull request merged**